src/utils/mutate.js

```diff
@@ -0,0 +1,242 @@
+import { chunk, cloneDeep, flatten, isFunction, mapValues } from 'lodash';
+import debug from 'debug';
+import { firestoreRef } from './query';
+import mark from './profiling';
+
+const info = debug('rrf:mutate');
+
+/**
+ * @param {object} firestore
+ * @param {string} collection
+ * @param {string} doc
+ * @returns Boolean
+ */
+const docRef = (firestore, collection, doc) =>
+  firestore.collection(collection).doc(doc);
+
+/**
+ * @param object
+ * @returns Promise
+ */
+async function promiseAllObject(object) {
+  return Object.fromEntries(
+    await Promise.all(
+      Object.entries(object).map(([key, promise]) =>
+        promise.then((value) => [key, value]),
+      ),
+    ),
+  );
+}
+
+/**
+ * @param {object} operations
+ * @returns Boolean
+ */
+function isBatchedWrite(operations) {
+  return Array.isArray(operations);
+}
+
+/**
+ * @param {Mutation_v1 | Mutation_v2} read
+ * @returns Boolean
+ */
+function isDocRead(read) {
+  return read && typeof read.doc === 'string';
+}
+
+/**
+ * @param {Mutation_v1 | Mutation_v2} operations
+ * @returns Boolean
+ */
+function isSingleWrite(operations) {
+  return operations && typeof operations.collection === 'string';
+}
+
+// ----- FieldValue support -----
+
+const primaryValue = (arr) =>
```

Comment:
Can these also be named functions to match the other utils? That way the could in theory be used within each other without an issue of ordering one util above/below another

Reply:
The mutate's `atomize` turns JSON data into Firestore writable format. The cacheReducer's `atomize` turns JSON into optimistic update JSON. 

PrimaryValue does need to go first. It decides if it's a custom tuple or not. The other just fall through until one is hit. 

In terms of scope they should only be in the `atomize` function. Assigning the vars outside the function was just to reduce recreating them on every pass. It could also be an inline switch but I thought this was easier to read. 

What do you think is the best place for this code?